Lib/urllib/request.py

```diff
@@ -350,6 +350,14 @@ def full_url(self):
     def full_url(self, url):
         # unwrap('<URL:type://host/path>') --> 'type://host/path'
         self._full_url = _unwrap(url)
+        # Sanity check self._full_url to avoid control characters in HTTP.
+        #  https://bugs.python.org/issue14826
+        #  https://bugs.python.org/issue36276
+        # The same control characters check was adopted by Golang in:
+        #  https://go-review.googlesource.com/c/go/+/159157
+        if (self._full_url.startswith('http') and
+            re.search("[\x00- \x7f-\x9f]", self._full_url)):
```

Comment:
As per fix in golang this should not include space as CTL and there is a test case where space is counted as a valid. So perhaps the regex could be `[\x00-\x19\x7f-\x9f]` (\x20 for space).

```golang
func isCTL(r rune) bool {
        return r < ' ' || 0x7f <= r && r <= 0x9f
}
```

Sample test case failure with url `http://a b:c d@python.org/` for current regex being used in `_splithost`

```
======================================================================
ERROR: test_userpass_inurl_w_spaces (test.test_urllib.urlopen_HttpTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/test/test_urllib.py", line 465, in test_userpass_inurl_w_spaces
    fp = urlopen(url)
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/test/test_urllib.py", line 47, in urlopen
    return opener.open(url)
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/urllib/request.py", line 1735, in open
    urltype, url = _splittype(fullurl)
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/urllib/parse.py", line 1003, in _splittype
    raise ValueError("URL can't contain control characters. %r" % (unquoted_url,))
ValueError: URL can't contain control characters. 'http://a b:c d@python.org/'

----------------------------------------------------------------------
Ran 95 tests in 0.650s
```

Reply:
note: `\x1f` is the char before space `\x20` not `\x19` :)

Reply:
Sorry, got little confused during conversion `hex(32) # 0x20` and used `\x19` in regex.